tbdex/rfq_create.go

```diff
@@ -0,0 +1,226 @@
+package tbdex
+
+import (
+	"crypto/rand"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	"go.jetpack.io/typeid"
+)
+
+// CreateRFQ creates an [RFQ]
+//
+// # An RFQ is a resource created by a customer of the PFI to request a quote
+//
+// [RFQ]: https://github.com/TBD54566975/tbdex/tree/main/specs/protocol#rfq-request-for-quote
+func CreateRFQ(from, to, offeringID string, payin PayinMethodWithPrivate, payout PayoutMethodWithPrivate, opts ...CreateRFQOption) (RFQ, error) {
+	defaultID, err := typeid.WithPrefix(RFQKind)
+	if err != nil {
+		return RFQ{}, fmt.Errorf("failed to generate default id: %w", err)
+	}
+
+	r := createRFQOptions{
+		id:         defaultID.String(),
+		createdAt:  time.Now(),
+		protocol:   "1.0",
+		externalID: "",
+	}
+
+	for _, opt := range opts {
+		opt(&r)
+	}
+
+	hashedData, privateData, err := hashData(&payin, &payout, r.claims)
+    if err != nil {
+		return RFQ{}, fmt.Errorf("failed to hash private data: %w", err)
+	}
+
+	return RFQ{
+		MessageMetadata: MessageMetadata{
+			From:       from,
+			To:         to,
+			Kind:       RFQKind,
+			ID:         r.id,
+			ExchangeID: r.id,
+			CreatedAt:  r.createdAt.UTC().Format(time.RFC3339),
+			ExternalID: r.externalID,
+			Protocol:   r.protocol,
+		},
+		Data: RFQData{
+			OfferingID: offeringID,
+			Payin:      SelectedPayinMethod{
+                Amount: payin.Amount,
+                Kind: payin.Kind,
+                PaymentDetailsHash: hashedData.PayinHash,
+            },
+			Payout:     SelectedPayoutMethod{
+                Kind: payin.Kind,
+                PaymentDetailsHash: hashedData.PayoutHash,
+            },
+			ClaimsHash: hashedData.ClaimsHash,
+		},
+		PrivateData: privateData,
+	}, nil
+}
+
+func hashData(payin *PayinMethodWithPrivate, payout *PayoutMethodWithPrivate, claims []string) (rfqHashes, RFQPrivateData, error) {
+    randomBytes := make([]byte, 16)
+	_, err := rand.Read(randomBytes)
+	if err != nil {
+		return rfqHashes{}, RFQPrivateData{}, err
+	}
```

Comment:
`web5-go` has a function for this [here](https://github.com/TBD54566975/web5-go/blob/main/crypto/entropy.go#L22)

Reply:
dopee